File: C#/3. Circular-Queue/CircularQueue/CircularQueue.cs
using System;

public class CircularQueue<T>
{
    private const int DEFAULT_CAPACITY = 16;
    private int head;
    private int tail;
    private T[] elements;

    public int Count { get; private set; }

    public CircularQueue(int capacity = DEFAULT_CAPACITY)
    {
        this.elements = new T[capacity];
    }

    public void Enqueue(T element)
    {
        if (this.Count >= this.elements.Length)
        {
            this.Grow();
        }
        this.elements[this.tail] = element;
        this.tail = (this.tail + 1) % this.elements.Length;
        this.Count++;
    }

    public T Dequeue()
    {
        if (this.Count == 0)
        {
            throw new InvalidOperationException("Queue is empty");
        }
        var result = this.elements[this.head];
        this.head = (this.head + 1) % this.elements.Length;
        this.Count--;
        return result;

    }

    public T[] ToArray()
    {
        T[] result = new T[this.Count];
        this.CopyElements(result);
        return result;
    }

    private void Grow()
    {
        T[] result = new T[2 * this.elements.Length];
        this.CopyElements(result);
        this.elements = result;
        this.head = 0;
        this.tail = this.Count;
    }

    private void CopyElements(T[] arr)
    {
        int index = this.head;
        for (int i = 0; i < this.Count; i++)
        {
            arr[i] = this.elements[index];
            index = (index + 1) % this.elements.Length;
        }
    }
}


class Example
{
    static void Main()
    {
        var queue = new CircularQueue<int>();

        queue.Enqueue(1);
        queue.Enqueue(2);
        queue.Enqueue(3);
        queue.Enqueue(4);
        queue.Enqueue(5);
        queue.Enqueue(6);

        Console.WriteLine("Count = {0}", queue.Count);
        Console.WriteLine(string.Join(", ", queue.ToArray()));
        Console.WriteLine("---------------------------");

        var first = queue.Dequeue();
        Console.WriteLine("First = {0}", first);
        Console.WriteLine("Count = {0}", queue.Count);
        Console.WriteLine(string.Join(", ", queue.ToArray()));
        Console.WriteLine("---------------------------");

        queue.Enqueue(-7);
        queue.Enqueue(-8);
        queue.Enqueue(-9);
        Console.WriteLine("Count = {0}", queue.Count);
        Console.WriteLine(string.Join(", ", queue.ToArray()));
        Console.WriteLine("---------------------------");

        first = queue.Dequeue();
        Console.WriteLine("First = {0}", first);
        Console.WriteLine("Count = {0}", queue.Count);
        Console.WriteLine(string.Join(", ", queue.ToArray()));
        Console.WriteLine("---------------------------");

        queue.Enqueue(-10);
        Console.WriteLine("Count = {0}", queue.Count);
        Console.WriteLine(string.Join(", ", queue.ToArray()));
        Console.WriteLine("---------------------------");

        first = queue.Dequeue();
        Console.WriteLine("First = {0}", first);
        Console.WriteLine("Count = {0}", queue.Count);
        Console.WriteLine(string.Join(", ", queue.ToArray()));
        Console.WriteLine("---------------------------");
    }
}

File: Java/src/circularQueue/CircularQueue.java
package circularQueue;

public class CircularQueue<E> {

	private final int DEFAULT_CAPACITY = 16;
	private int start;
	private int end;
	private E[] elements;
	private int count;

	@SuppressWarnings("unchecked")
	public CircularQueue(int capacity) {
		this.elements = (E[]) new Object[capacity];
	}

	@SuppressWarnings("unchecked")
	public CircularQueue() {
		this.elements = (E[]) new Object[DEFAULT_CAPACITY];
	}

	public int getCount() {
		return this.count;
	}

	public void enqueue(E element) {
		if (this.count >= this.elements.length) {
			this.grow();
		}

		this.elements[this.end] = element;
		this.end = (this.end + 1) % this.elements.length;
		this.count++;
	}

	public E dequeue() {
		if (this.count == 0) {
			throw new Error("Queue is empty");
		}
		E result = this.elements[this.start];
		this.start = (this.start + 1) % this.elements.length;
		this.count--;

		return result;
	}

	private void grow() {
		@SuppressWarnings("unchecked")
		E[] newElements = (E[]) new Object[2 * this.elements.length];

		this.copyElements(newElements);
		this.elements = newElements;
		this.start = 0;
		this.end = this.count;
	}

	private void copyElements(E[] newElements) {
		int index = this.start;
		for (int i = 0; i < this.count; i++) {
			newElements[i] = this.elements[index];
			index = (index + 1) % this.elements.length;
		}
	}

}
